File: backend/src/main/java/com/expensesage/service/CommentService.java
package com.expensesage.service;

import java.util.List;

import com.expensesage.dto.CommentCreateRequest;
import com.expensesage.dto.CommentResponseDto;
import com.expensesage.model.User;

public interface CommentService {

    /**
     * Retrieves all comments for a specific expense, ordered by creation time.
     * Ensures the requesting user is a member of the expense's group.
     *
     * @param expenseId   The ID of the expense.
     * @param currentUser The user requesting the comments.
     * @return A list of CommentResponseDto objects.
     * @throws RuntimeException if expense not found or user is not authorized.
     */
    List<CommentResponseDto> getCommentsForExpense(Long expenseId, User currentUser);

    /**
     * Adds a new comment to an expense.
     * Ensures the requesting user is a member of the expense's group.
     *
     * @param expenseId      The ID of the expense to comment on.
     * @param commentRequest The DTO containing the comment content.
     * @param currentUser    The user adding the comment.
     * @return The created CommentResponseDto.
     * @throws RuntimeException if expense not found or user is not authorized.
     */
    CommentResponseDto addCommentToExpense(Long expenseId, CommentCreateRequest commentRequest, User currentUser);

    /**
     * Deletes a comment.
     * Ensures the requesting user is the author of the comment.
     *
     * @param commentId   The ID of the comment to delete.
     * @param currentUser The user attempting to delete the comment.
     * @throws RuntimeException if comment not found or user is not the author.
     */
    void deleteComment(Long commentId, User currentUser);

}

File: frontend/src/services/commentService.ts
import apiClient from './api';
import { CommentCreateRequest, CommentResponseDto } from '../types/api';

/**
 * Fetches all comments for a specific expense.
 * @param expenseId The ID of the expense.
 * @returns A promise that resolves to an array of comments.
 */
export const getCommentsForExpense = async (expenseId: number): Promise<CommentResponseDto[]> => {
    try {
        const response = await apiClient.get<CommentResponseDto[]>(`/expenses/${expenseId}/comments`);
        return response.data;
    } catch (error: any) {
        console.error(`Failed to fetch comments for expense ${expenseId}:`, error.response?.data || error.message);
        throw error.response?.data || new Error('Failed to fetch comments');
    }
};

/**
 * Adds a new comment to an expense.
 * @param expenseId The ID of the expense to comment on.
 * @param commentData The comment content.
 * @returns A promise that resolves to the newly created comment.
 */
export const addCommentToExpense = async (expenseId: number, commentData: CommentCreateRequest): Promise<CommentResponseDto> => {
    try {
        const response = await apiClient.post<CommentResponseDto>(`/expenses/${expenseId}/comments`, commentData);
        return response.data;
    } catch (error: any) {
        console.error(`Failed to add comment to expense ${expenseId}:`, error.response?.data || error.message);
        throw error.response?.data || new Error('Failed to add comment');
    }
};

/**
 * Deletes a specific comment.
 * @param commentId The ID of the comment to delete.
 * @returns A promise that resolves when the deletion is successful.
 */
export const deleteComment = async (commentId: number): Promise<void> => {
    try {
        await apiClient.delete(`/comments/${commentId}`);
    } catch (error: any) {
        console.error(`Failed to delete comment ${commentId}:`, error.response?.data || error.message);
        throw error.response?.data || new Error('Failed to delete comment');
    }
};